compiler/hash-parser/src/lexer.rs

```diff
@@ -0,0 +1,618 @@
+                //                      stream each time we hit a Delimeter that uses some kind of brace. This could be very
+                //                      beneficial because we could parellilise the transformation of tokens into ASTs simply
+                //                      by observing the size and position of these children streams of tokens...
+                // >---------------------------------<
+                '(' => break TokenKind::OpenParen,
+                ')' => break TokenKind::CloseParen,
+                '{' => break TokenKind::OpenBrace,
+                '}' => break TokenKind::CloseBrace,
+                '[' => break TokenKind::OpenBracket,
+                ']' => break TokenKind::CloseBracket,
+                // >---------------------------------<
+                '~' => break TokenKind::Tilde,
+                '=' => break TokenKind::Eq,
+                '!' => break TokenKind::Exclamation,
+                '<' => break TokenKind::Lt,
+                '>' => break TokenKind::Gt,
+                '-' => break TokenKind::Minus,
+                '&' => break TokenKind::And,
+                '|' => break TokenKind::Pipe,
+                '+' => break TokenKind::Plus,
+                '*' => break TokenKind::Star,
+                '^' => break TokenKind::Caret,
+                '%' => break TokenKind::Percent,
+
+                // @@Improvement: This can be a potentially made out as a compound token... the same could be done with other
+                //                tokens that are likely to be compound, it would also avoid doing the work later on...
+                // ':' => match self.peek() {
+                //     ':' => {
+                //         self.next();
+                //         break TokenKind::NameAccess
+                //     }
+                //     _ => break TokenKind::Colon
+                // },
+                ':' => break TokenKind::Colon,
+
+                // Identifier (this should be checked after other variant that can
+                // start as identifier).
+                c if is_id_start(c) => break self.ident(),
+
+                // Numeric literal.
+                '0'..='9' => {
+                    break self
+                        .number()
+                        .unwrap_or_else(|e| panic!("error: {:#?}", e.message))
+                }
+                // character literal.
+                '\'' => {
+                    break self
+                        .char()
+                        .unwrap_or_else(|e| panic!("error: {:#?}", e.message)); // @@ErrorReporting: this is where we hook into error reporting to print the result
+                                                                                // and display it to the user
+                }
+
+                // String literal.
+                '"' => break self.string().unwrap_or_else(|e| panic!("error: {:?}", e)),
+
+                _ => break TokenKind::Unexpected,
```

Comment:
There should be an unexpected token kind?

Reply:
It's added to enhance error handling and reporting capabilities. When these tokens are converted into AST nodes, we might want to report the error and then try and recover to see if there are more errors. A good example of where this behaviour could be useful in handling the `Greek question-mark tragedy`. There might be a greek question mark instead of a semi-colon, which is unexpected and thus we might want to report it with additional context.